## Review: troubleshooting & advanced-usage guides

Great PR — the scope is well-targeted and every API claim I verified against the source code is accurate. The two new pages fill genuine gaps. Below are issues and suggestions, grouped by severity.

---

### Blockers

**README links use relative `.md` paths that break outside GitHub**

The new links in README.md point to `docs/docs/getting-started/advanced-usage.md`. These render fine on GitHub, but the README is the first thing users see on **PyPI** (`pip install` page) and in many editors — where the relative path resolves to nothing. The existing `CONTRIBUTING.md` and notebook links follow the same pattern, so this is a pre-existing issue, but the new links multiply the exposure.

**Recommendation:** Use absolute links to the docs site (`https://frgfm.github.io/torch-cam/getting-started/advanced-usage.html`) for the README, or at minimum add a note that they work on GitHub but not on PyPI. For `docs/docs/index.md` (which only lives on the docs site), relative paths are fine.

---

### High — should fix before merge

**1. `normalized` parameter undocumented**

`advanced-usage.md` shows the call signature `cam_extractor(class_idx, scores=None, normalized=True)` but never explains what `normalized` controls. For many users this is the first place they encounter it — it should say whether `False` returns raw (unnormalized) maps and when that's useful (e.g. comparing across layers before fusion).

**2. `fuse_cams` example uses undefined `class_idx`**

The LayerCAM list-of-layers example uses `class_idx` without defining it. Since this is a standalone code block in an "advanced usage" guide (not inside a `with GradCAM(model) as cam_extractor:` context), the reader has to mentally reconstruct the full context. Either wrap it in a complete runnable snippet or add a one-line comment like `# class_idx = out.argmax(dim=1).item()`.

**3. Multi-input wrapper: `target_layer` name is architecture-dependent**

The `LogitsOnly` wrapper example hardcodes `target_layer="model.backbone.layer4"`. The actual prefix depends on the specific model's `named_modules()` output — a user with a different architecture will copy this and get a key error. Add a note like `# check the actual name with: [n for n, _ in wrapped.named_modules() if 'layer4' in n][0]` or use `target_layer=wrapped.model.backbone.layer4` (module reference) which avoids the naming fragility entirely.

**4. `torch.no_grad()` during inference is not addressed**

Many production inference pipelines wrap everything in `torch.no_grad()`. The troubleshooting page correctly identifies the problem but doesn't mention this common pattern — a user running `with torch.no_grad(): model(input)` for efficiency will hit the same error. A short note like *"If your inference pipeline uses `torch.no_grad()`, move the CAM extraction outside that context manager"* would close this gap.

---

### Medium — worth addressing

**5. ViT section is too vague to act on**

The "Vision Transformers" section says *"reshape a block's token output back to a spatial grid ... with a small wrapper module"* but doesn't show even a skeleton example. Given that ViT questions are the 6th most common topic in issues/discussions, a 5-line wrapper snippet would be high-ROI. If you don't want to commit to a full example here, at least link to the relevant discussion thread.

**6. `retain_graph` is only shown for 2 classes**

The troubleshooting section shows the 2-class case, but users commonly need to compare 3+ classes. A brief note that `retain_graph=True` must be passed on *every* call but the last (not just the second) would prevent a wave of follow-up questions.

**7. Typo: `AssertionError` in troubleshooting heading**

The heading reads `` `AssertionError: Inputs need to be forwarded in the model ...` ``. This matches the source code typo (`core.py:163`), but the canonical English spelling is "AssertionError". Since this is user-facing documentation (not a stack trace), use the correct spelling and note the source typo in an italic aside. This helps users who search for the error.

**8. Missing cross-links between pages**

Troubleshooting mentions "try a different target_layer" but doesn't link to the target-layer table in advanced-usage. Similarly, advanced-usage doesn't link back to troubleshooting for the common errors a new user will hit. A few `[see Choosing the target layer](advanced-usage.md#choosing-a-target-layer)` and `[see Troubleshooting](troubleshooting.md#the-heatmap-is-blank--all-zeros)` links would make the docs self-reinforcing.

**9. `__all__` methods list incomplete in gradient-method table**

The "Choosing a CAM method" table lists `GradCAM`, `GradCAMpp`, `XGradCAM`, `SmoothGradCAMpp`, and `LayerCAM` as gradient-based, but omits `SmoothGradCAMpp` from the activation-based table while including it in the gradient table — this is correct, but the table could also note that `SmoothGradCAMpp` actually *does* use gradients internally (it's not activation-based). The current table already says "Needs gradients: yes" which is correct, so this is fine as-is. Just noting it.

---

### Low — suggestions for improvement

**10. `input_shape` default should link to auto-detection**

The `input_shape` section says the default is `(3, 224, 224)`. Worth a one-liner noting that omitting `target_layer` triggers auto-detection via the dummy forward — and that the log message will tell you which layer was picked (so users know to check `logging` output).

**11. Consider a "quick reference" callout box**

The README already has a good `[!TIP]` callout. Consider adding a collapsible "Quick reference" section at the top of advanced-usage that summarizes the 3 most common next steps (custom model, target layer, batched input) with 1-line code snippets. This serves users who want an answer in 10 seconds without reading the full page.

**12. 3D visualization snippet could use a note on dims**

The 3D example does `cam[0, 32].cpu().numpy()` — for users unfamiliar with volumetric data, a comment like `# cam shape: (N, D, H, W) — index depth dimension 32` would help.

---

### Summary

This is a well-structured, well-researched PR. The two high-impact fixes are: (1) the README links that won't resolve on PyPI, and (2) adding the missing `normalized` parameter explanation. The rest are incremental quality improvements that would make the docs more self-sufficient and reduce future support load.

Happy to help iterate on any of these.

--- a/README.md
+++ b/README.md
@@ -122,9 +122,9 @@ plt.imshow(result); plt.axis('off'); plt.tight_layout(); plt.show()
 ![overlayed_heatmap](https://github.com/frgfm/torch-cam/releases/download/v0.1.2/overlayed_heatmap.png)
 
 > [!TIP]
-> Using your own (non-torchvision) model, a Vision Transformer, 3D/video data, or batched inputs? Read the [**Advanced usage guide**](docs/docs/getting-started/advanced-usage.md) — it also covers how to choose the right `target_layer` and CAM method.
+> Using your own (non-torchvision) model, a Vision Transformer, 3D/video data, or batched inputs? Read the [**Advanced usage guide**](https://frgfm.github.io/torch-cam/getting-started/advanced-usage/) — it also covers how to choose the right `target_layer` and CAM method.
 >
-> Hitting a `cannot register a hook ...` / `requires grad` error, a `NaN`, or a blank heatmap? See [**Troubleshooting**](docs/docs/getting-started/troubleshooting.md).
+> Hitting a `cannot register a hook ...` / `requires grad` error, a `NaN`, or a blank heatmap? See [**Troubleshooting**](https://frgfm.github.io/torch-cam/getting-started/troubleshooting/).
 
 ## Setup
 
@@ -161,7 +161,7 @@ This project is developed and maintained by the repo owner, but the implementati
 - [XGrad-CAM](https://arxiv.org/abs/2008.02312): improved version of Grad-CAM in terms of sensitivity and conservation.
 - [Layer-CAM](http://mftp.mmcheng.net/Papers/21TIP_LayerCAM.pdf): Grad-CAM alternative leveraging pixel-wise contribution of the gradient to the activation.
 
-*Not sure which one to use? See [Choosing a CAM method](docs/docs/getting-started/advanced-usage.md#choosing-a-cam-method).*
+*Not sure which one to use? See [Choosing a CAM method](https://frgfm.github.io/torch-cam/getting-started/advanced-usage/#choosing-a-cam-method).*
 
 <p align="center">
     <a alt="wallaby_video_cam">

--- a/docs/docs/getting-started/advanced-usage.md
+++ b/docs/docs/getting-started/advanced-usage.md
@@ -1,7 +1,8 @@
 # Advanced usage
 
 The [quick start](../index.md) uses a torchvision classifier, but TorchCAM works with any PyTorch model. This
-guide covers the questions that come up most often once you move past the basic example.
+guide covers the questions that come up most often once you move past the basic example. Hitting an error rather
+than a usage question? Jump to [Troubleshooting](troubleshooting.md).
 
 ## Use your own model
 
@@ -41,24 +42,33 @@ A CAM is computed on the activation map of a **convolutional (spatial)** layer. 
 convolutional layer before global pooling — is the most class-discriminative but also the coarsest. Earlier
 layers give finer, less semantic maps. Rules of thumb for common torchvision backbones:
 
-| Architecture        | Typical `target_layer`   | `fc_layer` (only for `CAM`)        |
-| ------------------- | ------------------------ | ---------------------------------- |
-| ResNet / ResNeXt    | `"layer4"`               | `"fc"`                             |
-| DenseNet            | `"features"`             | `"classifier"`                     |
-| MobileNet v2 / v3   | `"features"`             | *multiple FC — `CAM` unsupported*  |
-| EfficientNet        | `"features"`             | *multiple FC — `CAM` unsupported*  |
-| SqueezeNet          | `"features"`             | *no FC — `CAM` unsupported*        |
-| VGG                 | `"features"`             | *multiple FC — `CAM` unsupported*  |
+| Architecture          | Typical `target_layer`   | `fc_layer` for `CAM`                 |
+| --------------------- | ------------------------ | ------------------------------------ |
+| ResNet / ResNeXt      | `"layer4"`               | `"fc"`                               |
+| DenseNet              | `"features"`             | `"classifier"`                       |
+| MobileNet v2          | `"features"`             | `"classifier.1"`                     |
+| EfficientNet          | `"features"`             | `"classifier.1"`                     |
+| MobileNet v3          | `"features"`             | *two `Linear` layers — `CAM` n/a*    |
+| VGG                   | `"features"`             | *three `Linear` layers — `CAM` n/a*  |
+| SqueezeNet            | `"features"`             | *no `Linear` head — `CAM` n/a*       |
+
+!!! note "When does the base `CAM` work?"
+    `CAM` needs **exactly one** `nn.Linear` classification head fed by global pooling, and resolves it
+    automatically. It therefore works for ResNet, DenseNet, MobileNet v2, EfficientNet, etc., but **not** for
+    models with several linear layers (VGG, MobileNet v3) or none (SqueezeNet) — there, use a gradient- or
+    score-based method, or pass a compatible `fc_layer` explicitly. All the other methods have no such
+    requirement.
 
 You can also pass a **list** of layers and fuse them — `LayerCAM` benefits a lot from this:
 
 ```python
 from torchcam.methods import LayerCAM
 
-cam_extractor = LayerCAM(model, ["layer2", "layer3", "layer4"])
-out = model(input_tensor)
-cams = cam_extractor(class_idx, out)        # one map per layer
-fused = cam_extractor.fuse_cams(cams)       # single fused map
+with LayerCAM(model, ["layer2", "layer3", "layer4"]) as cam_extractor:
+    out = model(input_tensor)
+    class_idx = out.squeeze(0).argmax().item()
+    cams = cam_extractor(class_idx, out)        # one map per layer
+    fused = cam_extractor.fuse_cams(cams)       # single fused map
 ```
 
 ## Understanding `class_idx` and the call signature
@@ -72,8 +82,15 @@ cam_extractor(class_idx, scores=None, normalized=True)
   index to see where the model looks for that class. For a batch, pass one index per sample (see below).
 - **`scores`** — the raw model output of shape `(N, num_classes)`. Required by the gradient-based methods (used
   for backprop) and by the Score-CAM family; ignored by `SmoothGradCAMpp` and `CAM`.
+- **`normalized`** — when `True` (default) each map is min-max normalized to `[0, 1]`, which is what you want for
+  visualization/overlay. Pass `normalized=False` to get the raw weighted maps, e.g. when comparing magnitudes
+  across layers before fusing them yourself.
 - **Returns** a `list` of activation maps, **one tensor per hooked layer**, each of shape `(N, H, W)`. With a
   single layer and a single image, the map you want is `cams[0].squeeze(0)`.
+
+Gradient-based extractors also accept `retain_graph=True` (forwarded to `loss.backward`), needed when you call
+the extractor several times after a single forward — see
+[Troubleshooting](troubleshooting.md#runtimeerror-trying-to-backward-through-the-graph-a-second-time).
 
 ## Batched inputs
 
@@ -109,8 +126,16 @@ class LogitsOnly(nn.Module):
         return self.model(x)[0]          # keep the logits, drop the rest
 
 wrapped = LogitsOnly(model)
-# target_layer names are now prefixed by "model."
-cam_extractor = GradCAM(wrapped, target_layer="model.backbone.layer4")
+cam_extractor = GradCAM(wrapped, target_layer=wrapped.model.backbone.layer4)  # pass the module directly
+```
+
+Passing the **module object** (rather than its name) sidesteps the naming gotcha: wrapping shifts every layer
+name under a `"model."` prefix, so a hard-coded string like `"backbone.layer4"` would raise a `ValueError`. If
+you prefer names, discover the correct one *after* wrapping:
+
+```python
+print([n for n, _ in wrapped.named_modules() if n.endswith("layer4")])
+# -> ['model.backbone.layer4']
 ```
 
 ## Vision Transformers and other non-CNN models
@@ -119,15 +144,37 @@ TorchCAM's methods operate on **convolutional feature maps** of shape `(N, C, H,
 3D). Transformer blocks emit token sequences of shape `(N, num_tokens, dim)`, which have no spatial grid, so the
 CAM methods do not apply directly and automatic `target_layer` resolution will not find a suitable layer.
 
-To use a ViT you have to reshape a block's token output back to a spatial grid (dropping the class token) before
-TorchCAM reads it — e.g. with a small wrapper module that reshapes `(N, num_tokens, dim)` to `(N, dim, H, W)` and
-expose *that* as the `target_layer`. There is no built-in helper for this yet; if you need it, please chime in on
-the [discussions](https://github.com/frgfm/torch-cam/discussions).
+To use a ViT you have to reshape a block's token output back to a spatial grid (dropping the class token) and
+expose *that* as the `target_layer`. There is no built-in helper yet, but the wrapper below is a starting point —
+it turns a block's `(N, num_tokens, dim)` output into an `(N, dim, H, W)` map:
+
+```python
+import torch.nn as nn
+
+class ViTToGrid(nn.Module):
+    """Experimental: reshape a ViT block's token output to a spatial grid for CAM."""
+    def __init__(self, vit_block, h, w):   # h, w = patch grid, e.g. 14x14 for 224px / patch 16
+        super().__init__()
+        self.block = vit_block
+        self.h, self.w = h, w
+
+    def forward(self, x):
+        out = self.block(x)                                  # (N, num_tokens, dim)
+        return out[:, 1:].transpose(1, 2).reshape(x.size(0), -1, self.h, self.w)  # (N, dim, h, w)
+```
+
+Insert it into the model in place of the block you want to read, then point the extractor at it.
+
+!!! warning "Experimental"
+    This is a sketch, not a drop-in. The exact reshape is architecture-specific (patch grid size, and whether
+    there is a class and/or distillation token to drop), and for gradient-based methods you must read a block
+    whose tokens the class score actually depends on. If you get this working for a model, please share it in the
+    [discussions](https://github.com/frgfm/torch-cam/discussions).
 
 ## 3D and video models
 
-Volumetric inputs work out of the box: set `input_shape` to your 3D shape (excluding the batch dimension) and the
-resulting map has shape `(N, D, H, W)`. Visualize it slice by slice:
+Volumetric inputs work out of the box: set `input_shape` to your 3D input shape as `(C, D, H, W)` (i.e. excluding
+the batch dimension) and the resulting map has shape `(N, D, H, W)`. Visualize it slice by slice:
 
 ```python
 import matplotlib.pyplot as plt

--- a/docs/docs/getting-started/troubleshooting.md
+++ b/docs/docs/getting-started/troubleshooting.md
@@ -32,6 +32,15 @@ disabled during the forward — almost always because it ran inside `torch.no_gr
         cams = cam_extractor(out.squeeze(0).argmax().item(), out)
     ```
 
+If your inference pipeline wraps everything in `torch.no_grad()` / `torch.inference_mode()` for speed, move just
+the CAM extraction (the hooked forward + the extractor call) outside that context — the rest of your pipeline can
+stay under `no_grad` if you like.
+
+!!! note "`model.eval()` and `torch.no_grad()` are orthogonal"
+    `eval()` only switches dropout/batch-norm to inference behavior; it does **not** disable autograd, so it is
+    safe — and recommended — for every method. The thing to avoid for gradient-based methods is
+    `no_grad()`/`inference_mode()`, *not* `eval()`. You generally want `model.eval()` **and** autograd enabled.
+
 !!! tip "Activation-based methods don't need gradients"
     `CAM`, `ScoreCAM`, `SSCAM` and `ISCAM` do not backpropagate, so you *can* keep their forward pass inside
     `torch.inference_mode()`. Only the gradient-based methods require autograd.
@@ -64,8 +73,12 @@ Gradient methods apply a `ReLU` before normalization, so a layer with no positiv
 class produces an empty map. Common causes:
 
 - `class_idx` is not the class the layer responds to — pass the prediction: `out.squeeze(0).argmax().item()`.
-- The target layer is too shallow/deep — try `LayerCAM` (keeps pixel-wise positive contributions) or another layer.
+- The target layer is too shallow/deep — try `LayerCAM` (keeps pixel-wise positive contributions) or
+  [another layer](advanced-usage.md#choosing-the-target-layer).
 - The model is untrained or the input is out of distribution.
+- With the activation-based `CAM`, a `fc_layer` that doesn't match the chosen `target_layer` (its channel count
+  or the head it feeds) yields a meaningless or empty map — let `CAM` resolve `fc_layer` automatically, or pass a
+  compatible one (see [Choosing the target layer](advanced-usage.md#choosing-the-target-layer)).
 
 ## The CAM changes on every run
 

